File: HeightBalanced.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 15 11:51:28 2020

@author: kevin
"""
"""
 This problem was recently asked by Twitter:
Given a tree, find if the binary tree is height balanced or not. A height 
balanced binary tree is a tree where every node's 2 subtree do not differ in 
height by more than 1.
Here's some starter code:
"""
class Node:
  def __init__(self, value, left=None, right=None):
    self.value = value
    self.left = left
    self.right = right

def height(node):
    if(not node):
        return 0
    else:
        return 1 + max(height(node.left), height(node.right))

def is_height_balanced(tree):
    # Fill this in.
    if(not tree):
        return True
    if(abs(height(tree.left)-height(tree.right)) <= 1 and is_height_balanced(tree.left) 
    and is_height_balanced(tree.right)):
        return True
    return False

#     1
#    / \
#   2   3
#  /
# 4  
n4 = Node(4)
n3 = Node(3)
n2 = Node(2, n4)
n1 = Node(1, n2, n3)

print(is_height_balanced(n1))
# True

#     1
#    / 
#   2   
#  /
# 4  
n1 = Node(1, n2)
print(is_height_balanced(n1))
# False

File: HeightBalanced.java
/*
This problem was recently asked by Twitter:
Given a tree, find if the binary tree is height balanced or not. A height 
balanced binary tree is a tree where every node's 2 subtree do not differ in 
height by more than 1.
*/
import java.lang.Math;

class HeightBalanced{
   public static void main(String[] args){
      /*   1
          / \
         2   3
        /
       4  
      */
      TreeNode root = new TreeNode(1, new TreeNode(2, new TreeNode(4)), new TreeNode(3));
      System.out.println(root);
      System.out.println(heightBalanced(root)); // True
   }
   
   /**
   * Checks to see if a binary tree is height balanced
   *
   * @param root The root of the binary tree
   * @return True if the tree is height balanced, false otherwise
   */
   static boolean heightBalanced(TreeNode root){
      if(root != null){
         if(!(Math.abs(height(root.left)-height(root.right))<=1 && heightBalanced(root.left) && heightBalanced(root.right))){
            return false;
         }
      }
      return true;
   }
   
   /**
   * Gets the length of a binary tree
   *
   * @param root The root of the binary tree
   * @return The number of nodes in the longest path of a tree
   */
   static int height(TreeNode root){
      if(root == null){
         return 0;
      }
      return 1 + Math.max(height(root.left), height(root.right));
   }
}

class TreeNode{
   int data;
   TreeNode left;
   TreeNode right;
   
   public TreeNode(int data){
      this.data = data;
   }
   public TreeNode(int data, TreeNode left, TreeNode right){
      this.data = data;
      this.left = left;
      this.right = right;
   }
   public TreeNode(int data, TreeNode left){
      this.data = data;
      this.left = left;
   }
   
   public String toString(){
      return data + "(" + this.left + ", " + this.right + ")";
   }
}